core/src/main/scala/org/apache/spark/memory/MemoryManager.scala

```diff
@@ -255,7 +256,14 @@ private[spark] abstract class MemoryManager(
     }
     val size = ByteArrayMethods.nextPowerOf2(maxTungstenMemory / cores / safetyFactor)
     val default = math.min(maxPageSize, math.max(minPageSize, size))
-    conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    val sizeAsBytes = conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    if (Utils.isG1GarbageCollector &&
+      tungstenMemoryMode == MemoryMode.ON_HEAP &&
+      sizeAsBytes % (1024 * 1024) == 0) {
+      sizeAsBytes - Platform.LONG_ARRAY_OFFSET
```

Comment:
Could you update the function description with this new addition?

Reply:
updated